Makefile

```diff
@@ -217,7 +217,7 @@ tools: # Install dependencies and tools required to build
 
 .PHONY: test
 test: # Run tests
-	go test ./...
+	go test ./... -race
```

Comment:
Should we have `-race` be the default here, or separate that into it's own Make step? There is a potentially [non-trivial](https://go.dev/doc/articles/race_detector#Runtime_Overheads) overhead cost here, though I fully admit I don't know if it impacts us in practice. 

Looking at other projects like Vault and Consul, they break out the race test into their own make step and I wonder if we should follow suit. Since we're adding it to the CI step above, even if a developer fails to check the race test before opening a PR we should still be covered there. Thoughts?

Reply:
Link to the consul separate race step for anyone following along: https://github.com/hashicorp/consul/blob/8dee3534927eb50572b003e6335f65e063b36701/.github/workflows/go-tests.yml#L258-L276

I can't think of a compelling benefit for us to break the race test into a separate stage, and run them in parallel with the regular tests. Downsides would be complexity when looking at our CI run, double the chance for flaky tests to fail, and additional GHA cost.

> Since we're adding it to the CI step above, even if a developer fails to check the race test before opening a PR we should still be covered there. Thoughts?

I don't think I follow this! I agree that with this CI check, if a developer makes a PR that includes a test with a race condition, this should catch it as part of the test suite, but i'm not sure how that relates to making the race tests their own parallel CI step.

Reply:
> I can't think of a compelling benefit for us to break the race test into a separate stage, and run them in parallel with the regular tests. Downsides would be complexity when looking at our CI run

My suggestion/comment here is not regarding our CI run(s), it's just on the Makefile step `test`. To my knowledge our CI steps do not run `make test`, but I could be wrong. 

> but i'm not sure how that relates to making the race tests their own parallel CI step.

I considered making that suggestion but decided not to. The line I'm commenting on is just in our `Makefile` and not the `go test` CI step.

Reply:
I feel like the implicit promise with `make test` is that "if `make test` passes locally, it should also pass in CI". 

I recognize that we're not living up to that currently (CI ignores the e2e tests and `make tests` doesn't), but I feel like the goal should be to keep it as close to in sync with the CI tests as possible. If we make the race detector tests  a separate make file step, it introduces the potential for a contributor to be confused about why the tests are passing locally and failing in CI.

Reply:
> "if make test passes locally, it should also pass in CI"

Generally, yeah that makes sense, with the exception(s) of matrix tests/jobs that test different versions, but that's a technicality ... 

> If we make the race detector tests a separate make file step, it introduces the potential for a contributor to be confused about why the tests are passing locally and failing in CI.

That's fair and valid. If as you mention with Evan above the impact is minimal then I'm fine with `-race` being the default 👍